File: scripts/algorithms/F/Find Two Nonoverlapping Subarrays Each With Target Sum/Find Two Non-overlapping Sub-arrays Each With Target Sum.cpp
// Runtime: 278 ms (Top 54.50%) | Memory: 77.1 MB (Top 64.88%)
class Solution {
public:
    int minSumOfLengths(vector<int>& arr, int target) {
        int n=arr.size();
        vector<int> prefix(n,INT_MAX);
        vector<int> suffix(n,INT_MAX);
        int sum=0;
        int start=0;
        for(int end=0;end<n;end++){
            sum+=arr[end];
            while(sum>target){
                sum-=arr[start++];
            }
            if(sum==target){
                prefix[end]=min(prefix[end-1>=0 ? end-1 : 0],end-start+1);
            }
            else{
                prefix[end]=prefix[end-1>=0 ? end-1 : 0];
            }
        }

        sum=0;
        start=n-1;
        for(int end=n-1;end>=0;end--){
            sum+=arr[end];
            while(sum>target){
                sum-=arr[start--];
            }
            if(sum==target){
                suffix[end]=min(suffix[end+1<n ? end+1 : n-1],start-end+1);
            }
            else{
                suffix[end]=suffix[end+1<n ? end+1 : n-1];
            }
        }
        int res=INT_MAX;
        for(int i=0;i<n-1;i++){
            if(prefix[i]==INT_MAX || suffix[i+1]==INT_MAX){
                continue;
            }
            res=min(res,prefix[i]+suffix[i+1]);
        }
        return res==INT_MAX ? -1 : res;

    }
};

File: scripts/algorithms/F/Find Two Nonoverlapping Subarrays Each With Target Sum/Find Two Nonoverlapping Subarrays Each With Target Sum.java
// Runtime: 12 ms (Top 54.1%) | Memory: 57.97 MB (Top 80.3%)

class Solution {
    public int minSumOfLengths(int[] arr, int target) {
        int[] left = findMinLengths(0, arr.length, target,arr);
        int[] right = findMinLengths(arr.length-1, 0,target,arr);
        
        for(int i=1;i<arr.length;i++)
            left[i] = Math.min(left[i],left[i-1]);
        
        for(int i=arr.length-2;i>=0;i--)
            right[i] = Math.min(right[i],right[i+1]);
        
        int min = arr.length+1;
        
        for(int i=0;i<arr.length-1;i++)
            min = Math.min(min,left[i] + right[i+1]);
        
        return min == arr.length+1 ? -1 : min;
    }
    
    private int[] findMinLengths(int start,int end,int target,int[] arr){
        int[] ans = new int[arr.length];
        
        Arrays.fill(ans,arr.length);
        
        int i = start,j = start;
        int sum = 0;
        
        int dir = (int)Math.signum(end-start);
        
        while(j != end){
            sum += arr[j];
            
            while(sum > target){
                sum -= arr[i];
                i += dir;
            }
                
            if(sum == target){
                ans[j] = Math.abs(i-j)+1;
            }
            
            j+=dir;
        }
        
        return ans;
    }
}